btree/btree.c

```diff
@@ -3892,24 +3931,193 @@ static void m0_btree_ut_basic_kv_operations(void)
+						   &b_op.bo_op_exec);
 
 		for (i = 1; i < 2048; i++) {
 			search_key = key;
-			M0_BTREE_OP_SYNC_WITH(&kv_op.bo_op,
+			M0_BTREE_OP_SYNC_WITH_RC(&kv_op.bo_op,
 					      m0_btree_nxt(b_op.bo_arbor,
 							   &search_key_in_tree,
 							   &ut_cb, 0, &kv_op),
-					      &G, &b_op.bo_op_exec);
+							   &G,
+							   &b_op.bo_op_exec);
+		}
+	}
+
+	m0_btree_close(b_op.bo_arbor);
+	M0_BTREE_OP_SYNC_WITH_RC(&b_op.bo_op,
+			      m0_btree_destroy(b_op.bo_arbor, &b_op), &G,
+			      &b_op.bo_op_exec);
+	btree_ut_fini();
+}
+
+#define MIN_STREAM_CNT             5
+#define MAX_STREAM_CNT             20
+
+#define MIN_RECS_PER_STREAM        5
+#define MAX_RECS_PER_STREAM        2048
+
+/**
+ * This unit test exercises the KV operations triggered by multiple streams.
+ */
+static void m0_btree_ut_multi_stream_kv_oper(void)
+{
+	void                   *temp_node;
+	int                     i;
+	time_t                  curr_time;
+	struct m0_btree_cb      ut_cb;
+	struct m0_be_tx        *tx              = NULL;
+	struct m0_btree_op      b_op            = {};
+	uint32_t                stream_count    = 0;
+	uint64_t                recs_per_stream = 0;
+	struct m0_btree_op      kv_op           = {};
+	const struct node_type *nt              = &fixed_format;
+	struct m0_btree_type    btree_type      = {.tt_id = M0_BT_UT_KV_OPS,
+						  .ksize = sizeof(uint64_t),
+						  .vsize = btree_type.ksize*2,
+						  };
+
+	M0_ENTRY();
+
+	time(&curr_time);
+	printf("\nUsing seed %lu", curr_time);
+	srandom(curr_time);
+
+	stream_count = (random() % (MAX_STREAM_CNT - MIN_STREAM_CNT)) +
+			MIN_STREAM_CNT;
+
+	recs_per_stream = (random()%
+			   (MAX_RECS_PER_STREAM - MIN_RECS_PER_STREAM)) +
+			    MIN_RECS_PER_STREAM;
+
+	/** Prepare transaction to capture tree operations. */
+	m0_be_tx_init(tx, 0, NULL, NULL, NULL, NULL, NULL, NULL);
+	m0_be_tx_prep(tx, NULL);
+	btree_ut_init();
+	/**
+	 *  Run valid scenario:
+	 *  1) Create a btree
+	 *  2) Adds records in multiple streams to the created tree.
+	 *  3) Confirms the records are present in the tree.
+	 *  4) Deletes all the records from the tree using multiple streams.
+	 *  5) Close the btree
+	 *  6) Destroy the btree
+	 */
+
+	/** Create temp node space and use it as root node for btree */
+	temp_node = m0_alloc_aligned((1024 + sizeof(struct nd)), 10);
+	M0_BTREE_OP_SYNC_WITH_RC(&b_op.bo_op,
+			      m0_btree_create(temp_node, 1024, &btree_type, nt,
+					      tx, &b_op), &G, &b_op.bo_op_exec);
+
+	for (i = 1; i <= recs_per_stream; i++) {
+		uint64_t             key;
+		uint64_t             value[btree_type.vsize / sizeof(uint64_t)];
+		struct cb_data       put_data;
+		struct m0_btree_rec  rec;
+		m0_bcount_t          ksize  = sizeof key;
+		m0_bcount_t          vsize  = sizeof value;
+		void                *k_ptr  = &key;
+		void                *v_ptr  = &value;
+		uint32_t             stream_num;
+
+		for (stream_num = 0; stream_num < stream_count; stream_num++) {
+			int k;
```

Comment:
can you shift line 4018 to 4025 at line 2029 as new record for each put operation is required, as record can get updated by put operation, so better to use new record for each put operation

Reply:
A new record is always sent to the put routine. I am sorry I donot follow your comments.